File: pset6/credit/credit.py

def main():
    # Global variable declarations
    cardNumber = input("Enter your credit card number: ")
    firstCopy = int(cardNumber)
    secondCopy = int(cardNumber) / 10
    
    # Checks initial validity of card number
    checkValidity(cardNumber)

    # Calculates Luhn's Algorithm for card number
    calculateLuhn(firstCopy, secondCopy)
    
    # Declares if it's AMEX\n or MASTERCARD\n or VISA\n or INVALID\n,
    declare(cardNumber)


def checkValidity(n):
    
    if len(str(n)) < 13 or len(str(n)) > 16:
        print("INVALID")
        exit(1)
    else:
        return True

    
def calculateLuhn(x, y):    
    totalSum = 0
    # Luhn's Algorithm
    # Part 1: Alternate Additions
    while True:
        currentCount = int(x % 10)
        totalSum += currentCount
        x /= 100
        if x == 0:
            break
    
    # Part 2: Other Alternate Additions
    multiplyNum = 0
    while True:
        currentCount = int(y % 10)
        multiplyNum = currentCount * 2
        totalSum = totalSum + int((multiplyNum % 10)) + int((multiplyNum / 10))
        y /= 100
        if y == 0:
            break
    
    isValid = bool(totalSum % 10 == 0)

    
def declare(n):
    # American Express uses 15-digit numbers, American Express numbers start with 34 or 37
    # MasterCard uses 16-digit numbers, and MasterCard numbers start with 51, 52, 53, 54, or 55 
    # Visa numbers start with 4 Visa uses 13- and 16-digit
    
    # Finding the first digit of the cardNumber by dividing it until we get 1s place
    firstDigit = int(n)
    while (firstDigit >= 10):
        firstDigit = int(firstDigit / 10)
    
    # Finding the last two digits of the cardNumber by dividing it until we get 10s place
    twoDigit = int(n)
    while (twoDigit >= 100):
        twoDigit = int(twoDigit / 10)
    
    digitCount = len(str(n))

    if firstDigit == 4 and (digitCount == 13 or digitCount == 16):
        print("VISA")
    elif (twoDigit == 34 or twoDigit == 37) and digitCount == 15:
        print("AMEX")
    elif (twoDigit in range(51, 56)) and digitCount == 16:
        print("MASTERCARD")
    else:
        print("INVALID")
main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>
#include <string.h>

void validateCreditCard(long cardNumber);

int main(void)
{
    long cardNumber;

    // Get number
    cardNumber = get_long("Enter your credit card number: \n");
    validateCreditCard(cardNumber);

    return 0;
}

void validateCreditCard(long cardNumber)
{
    int firstDigit;
    int twoDigits;
    int totalSum = 0;
    int currentCount = 0;

    char cardNumberString[256];
    sprintf(cardNumberString, "%ld", cardNumber);
    int cardNumberLength = strlen(cardNumberString);

    if (cardNumberLength < 13 || cardNumberLength > 16)
    {
        printf("INVALID\n");
        return;
    }

    long originalNum = cardNumber;



    // Checksum
    long cardNumberCopy = cardNumber;

    // For the first alternate additions

    do
    {
        currentCount = cardNumberCopy % 10; // To get the last digit
        totalSum += currentCount; // Adds last digit to the total
        cardNumberCopy /= 100; // Iteration
    }
    while (cardNumberCopy != 0);

    // For the second alternate additions

    int multiplyNum = 0;

    cardNumberCopy = cardNumber / 10;
    do
    {
        currentCount = cardNumberCopy % 10;
        multiplyNum = currentCount * 2; //According to Luhn's algorithm, we have to double the number
        totalSum = totalSum + (multiplyNum % 10) + (multiplyNum / 10);
        cardNumberCopy /= 100;

    }
    while (cardNumberCopy != 0);


    // Checking validity
    bool isValid = totalSum % 10 == 0;


    // All American Express numbers start with 34 or 37;
    // most MasterCard numbers start with 51, 52, 53, 54, or 55
    // And all Visa numbers start with 4.

    //  To check this criteria:
    int countDigit = 0;
    do
    {
        cardNumber /= 10;
        countDigit++;
    }
    while (cardNumber > 0);


    long divideBy = pow(10, countDigit - 2);
    
    twoDigits = originalNum / divideBy;
    firstDigit = (twoDigits / 10);

    if (isValid)
    {

        if (firstDigit == 4 && (countDigit == 13 || countDigit == 16))
        {
            printf("VISA\n");
        }
        else if ((twoDigits == 34 || twoDigits == 37) && countDigit == 15)
        {
            printf("AMEX\n");
        }
        else if ((twoDigits > 50 && twoDigits < 56) && countDigit == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}